# INSAID Hiring Exercise

## Important: Kindly go through the instructions mentioned below.

- The Sheet is structured in **4 steps**:
    1. Understanding data and manipulation
    2. Data visualization
    3. Implementing Machine Learning models(Note: It should be more than 1 algorithm)
    4. Model Evaluation and concluding with the best of the model.
    
    
    

- Try to break the codes in the **simplest form** and use number of code block with **proper comments** to them
- We are providing **h** different dataset to choose from(Note: You need to select any one of the dataset from this sample sheet only)
- The **interview calls** will be made solely based on how good you apply the **concepts**.
- Good Luck! Happy Coding!

### Importing the data

In [174]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.offline as py
py.init_notebook_mode(connected=True)
import plotly.graph_objs as go

In [175]:
df=pd.read_csv("Churn.csv")

In [176]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [177]:
len(df)

7043

### Understanding the data

In [178]:
df.PaymentMethod.unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [179]:
print("Null Entries: ",  df.isnull().sum(axis = 0))
print ("\n Unique values :\n", df.nunique())

print("\nBlanks in Total Charges ",(df['TotalCharges'] == ' ').sum())
print("Blanks in tenure  ",(df['tenure'] == ' ').sum())
print("Blanks in MonthlyCharges ",(df['MonthlyCharges'] == ' ').sum())
 


Null Entries:  customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

 Unique values :
 customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges    

/home/prashant/.local/lib/python3.5/site-packages/pandas/core/ops.py:1649: FutureWarning:

elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison



In [59]:
df.head()

gender SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female            No     Yes         No       1           No   
1    Male            No      No         No      34          Yes   
2    Male            No      No         No       2          Yes   
3    Male            No      No         No      45           No   
4  Female            No      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

### Data Manipulation

In [193]:
## Rem
df['TotalCharges'] = df["TotalCharges"].replace(" ",np.nan)
df.dropna(inplace=True)

df["SeniorCitizen"] = df["SeniorCitizen"].replace({"Yes":1,"No":0})

In [194]:
#del df['customerID']
df.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines OnlineSecurity OnlineBackup DeviceProtection  ...  \
0  No phone service             No          Yes               No  ...   
1                No            Yes           No              Yes  ...   
2                No            Yes          Yes               No  ...   
3  No phone service            Yes           No              Yes  ...   
4                No             No           No               No  ...   

  StreamingMovies        Contract PaperlessBilling              PaymentMethod  \
0              No  Month-to-month              Yes           Electronic check   
1              No        One year               No               Mailed check   
2              No  Month-to-month              Yes               Mailed check   
3              No        One year               No  Bank transfer (automatic)   
4              No  Month-to-month              Yes           Electronic check   

  MonthlyCharges TotalCharges  Churn InternetService_DSL  \
0          29.85        29.85      0                   1   
1          56.95       1889.5      0                   1   
2          53.85       108.15      1                   1   
3          42.30      1840.75      0                   1   
4          70.70       151.65      1                   0   

   InternetService_Fiber optic  InternetService_No  
0                            0                   0  
1                            0                   0  
2                            0                   0  
3                            0                   0  
4                            1                   0  

[5 rows x 22 columns]

## Data Visualization

In [388]:
layout = go.Layout(
    title=go.layout.Title(
        text='Plot Title',
        xref='paper',
        x=0
    ),
    xaxis=go.layout.XAxis(
        title=go.layout.xaxis.Title(
            text='Number of people churned',
            font=dict(
                family='Courier New, monospace',
                size=18,
                color='#7f7f7f'
            )
        )
    ),
    yaxis=go.layout.YAxis(
        title=go.layout.yaxis.Title(
            text='Churns',
            font=dict(
                family='Courier New, monospace',
                size=18,
                color='#7f7f7f'
            )
        )
    )
)

lab = df["Churn"].value_counts().keys().tolist()
val = df["Churn"].value_counts().values.tolist()

data = [go.Bar(x=lab,y=val)]

fig = go.Figure(data=data, layout=layout)
py.iplot(fig)

Not_Churned=val[0]/len(df)*100
Churned=val[1]/len(df)*100
print("              Customers Not_Churned =",Not_Churned,"%              Customers Churned =",Churned,"%")

              Customers Not_Churned = 73.42150170648463 %              Customers Churned = 26.578498293515356 %


In [56]:
gen=df["gender"].value_counts()
male=gen[0]/len(df)*100
female=gen[1]/len(df)*100

In [57]:
layout = go.Layout(
    title=go.layout.Title(
        text='Plot Title',
        xref='paper',
        x=0
    ),
    xaxis=go.layout.XAxis(
        title=go.layout.xaxis.Title(
            text='Gender',
            font=dict(
                family='Courier New, monospace',
                size=18,
                color='#7f7f7f'
            )
        )
    ),
    yaxis=go.layout.YAxis(
        title=go.layout.yaxis.Title(
            text='Percentag',
            font=dict(
                family='Courier New, monospace',
                size=18,
                color='#7f7f7f'
            )
        )
    )
)

data = [go.Bar(x=["Male","Female"] ,y=[male,female])]

fig = go.Figure(data=data, layout=layout)
py.iplot(fig)



##                                 Gender ratio is almost equal

In [71]:
contract=df["Contract"].value_counts()
month=(contract[0]/len(df))*100
two_year= contract[1]/len(df)*100
one_year= contract[2]/len(df)*100

In [76]:
labels = ['Month to Month','Two Year','One Year']
values = [month,two_year,one_year]
colors = ['#FEBFB3', '#E1396C', '#96D38C', '#D0F9B1']

trace = go.Pie(labels=labels, values=values,
               hoverinfo='percent', textinfo='label', 
               textfont=dict(size=20),
               marker=dict(colors=colors, 
                           line=dict(color='#000000', width=2)))

py.iplot([trace], filename='styled_pie_chart')

## Most of the customers are on Month to Month Subscription

In [84]:
senior=df["SeniorCitizen"].value_counts()
senior_no=(senior[0]/len(df))*100
senior_yes= senior[1]/len(df)*100

In [97]:
labels = ['Young','Senior']
values = [senior_no,senior_yes]
colors = ['#96D38C', '#D0F9B1']

trace = go.Pie(labels=labels, values=values,
               hoverinfo='percent', textinfo='label', 
               textfont=dict(size=20),
               marker=dict(colors=colors, 
                           line=dict(color='#000000', width=2)))

py.iplot([trace], filename='styled_pie_chart')

#### Most of the customers are Young

### Implement Machine Learning Models

In [188]:
#To convert all Yes or No to 1 & 0
df['Churn'].replace(to_replace='Yes', value=1, inplace=True)
df['Churn'].replace(to_replace='No',  value=0, inplace=True)

In [183]:
df.drop(["customerID"],inplace=True, axis=1)

In [184]:
replace_cols = [ 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
                'TechSupport','StreamingTV', 'StreamingMovies']
for i in replace_cols : 
    df[i]  = df[i].replace({'No internet service' : 'No'})

In [189]:
df.head()

gender SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female            No     Yes         No       1           No   
1    Male            No      No         No      34          Yes   
2    Male            No      No         No       2          Yes   
3    Male            No      No         No      45           No   
4  Female            No      No         No       2          Yes   

      MultipleLines OnlineSecurity OnlineBackup DeviceProtection  ...  \
0  No phone service             No          Yes               No  ...   
1                No            Yes           No              Yes  ...   
2                No            Yes          Yes               No  ...   
3  No phone service            Yes           No              Yes  ...   
4                No             No           No               No  ...   

  StreamingMovies        Contract PaperlessBilling              PaymentMethod  \
0              No  Month-to-month              Yes           Electronic check   
1              No        One year               No               Mailed check   
2              No  Month-to-month              Yes               Mailed check   
3              No        One year               No  Bank transfer (automatic)   
4              No  Month-to-month              Yes           Electronic check   

  MonthlyCharges TotalCharges  Churn InternetService_DSL  \
0          29.85        29.85      0                   1   
1          56.95       1889.5      0                   1   
2          53.85       108.15      1                   1   
3          42.30      1840.75      0                   1   
4          70.70       151.65      1                   0   

   InternetService_Fiber optic  InternetService_No  
0                            0                   0  
1                            0                   0  
2                            0                   0  
3                            0                   0  
4                            1                   0  

[5 rows x 22 columns]

#### To convert all categorial data to dummy data

In [186]:
dfDummies = pd.get_dummies(df['InternetService'], prefix = 'InternetService')
df = pd.concat([df, dfDummies], axis=1)
df.drop(["InternetService"],inplace=True, axis=1)

In [196]:
dfDumm = pd.get_dummies(df['PaymentMethod'], prefix = 'PaymentMethod')
df = pd.concat([df, dfDumm], axis=1)
df.drop(["PaymentMethod"],inplace=True, axis=1)

In [198]:
dfDumm1 = pd.get_dummies(df['MultipleLines'], prefix = 'MultipleLines')
df = pd.concat([df, dfDumm1], axis=1)
df.drop(["MultipleLines"],inplace=True, axis=1)

In [237]:
dfDumm2 = pd.get_dummies(df['Contract'], prefix = 'Contract')
df = pd.concat([df, dfDumm2], axis=1)
df.drop(["Contract"],inplace=True, axis=1)

In [227]:
df["gender"] = df["gender"].replace({"Male":1,"Female":0})

In [225]:
cols=["gender","SeniorCitizen","Partner","Dependents","PhoneService","OnlineSecurity","OnlineBackup","DeviceProtection","TechSupport","StreamingTV","StreamingMovies","Contract","PaperlessBilling","TotalCharges"]
for i in cols:
    df[cols] = df[cols].replace({"Yes":1,"No":0})

In [228]:
# Finally All Data is Numerical
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  ...  \
0               0             1                 0            0  ...   
1               1             0                 1            0  ...   
2               1             1                 0            0  ...   
3               1             0                 1            1  ...   
4               0             0                 0            0  ...   

   InternetService_DSL  InternetService_Fiber optic InternetService_No  \
0                    1                            0                  0   
1                    1                            0                  0   
2                    1                            0                  0   
3                    1                            0                  0   
4                    0                            1                  0   

   PaymentMethod_Bank transfer (automatic)  \
0                                        0   
1                                        0   
2                                        0   
3                                        1   
4                                        0   

   PaymentMethod_Credit card (automatic) PaymentMethod_Electronic check  \
0                                      0                              1   
1                                      0                              0   
2                                      0                              0   
3                                      0                              0   
4                                      0                              1   

   PaymentMethod_Mailed check  MultipleLines_No  \
0                           0                 0   
1                           1                 1   
2                           1                 1   
3                           0                 0   
4                           0                 1   

   MultipleLines_No phone service  MultipleLines_Yes  
0                               1                  0  
1                               0                  0  
2                               0                  0  
3                               1                  0  
4                               0                  0  

[5 rows x 27 columns]

In [315]:
X=df
Y=df['Churn']
X=X.drop(["Churn"], axis=1)

In [316]:
X['TotalCharges']=X['TotalCharges'].astype(float)
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)

In [317]:
print(len(x_train))
print(len(y_train))
print(len(x_test))
print(len(y_test))

5625
5625
1407
1407


In [238]:
pd.set_option('display.max_columns', 30)
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  StreamingTV  \
0               0             1                 0            0            0   
1               1             0                 1            0            0   
2               1             1                 0            0            0   
3               1             0                 1            1            0   
4               0             0                 0            0            0   

   StreamingMovies  PaperlessBilling  MonthlyCharges TotalCharges  Churn  \
0                0                 1           29.85        29.85      0   
1                0                 0           56.95       1889.5      0   
2                0                 1           53.85       108.15      1   
3                0                 0           42.30      1840.75      0   
4                0                 1           70.70       151.65      1   

   InternetService_DSL  InternetService_Fiber optic  InternetService_No  \
0                    1                            0                   0   
1                    1                            0                   0   
2                    1                            0                   0   
3                    1                            0                   0   
4                    0                            1                   0   

   PaymentMethod_Bank transfer (automatic)  \
0                                        0   
1                                        0   
2                                        0   
3                                        1   
4                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               1   
1                                      0                               0   
2                                      0                               0   
3                                      0                               0   
4                                      0                               1   

   PaymentMethod_Mailed check  MultipleLines_No  \
0                           0                 0   
1                           1                 1   
2                           1                 1   
3                           0                 0   
4                           0                 1   

   MultipleLines_No phone service  MultipleLines_Yes  Contract_Month-to-month  \
0                               1                  0                        1   
1                               0                  0                        0   
2                               0                  0                        1   
3                               1                  0                        0   
4                               0                  0                        1   

   Contract_One year  Contract_Two year  
0                  0                  0  
1                  1                  0  
2                  0                  0  
3                  1                  0  
4                  0                  0

### Model Evaluation

###  Logistic Regression

In [242]:
from sklearn.linear_model import LogisticRegression
logisticReg = LogisticRegression()
logisticReg.fit(x_train, y_train)

/home/prashant/.local/lib/python3.5/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning:

Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.



LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [244]:
predictions = logisticReg.predict(x_test)

In [394]:
score = logisticReg.score(x_test, y_test)*100

In [395]:
score

80.59701492537313

In [255]:
import matplotlib.pyplot as plt
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

In [256]:
print ("\n Classification report : \n",classification_report(y_test,predictions))


 Classification report : 
               precision    recall  f1-score   support

           0       0.85      0.90      0.87      1038
           1       0.66      0.54      0.59       369

    accuracy                           0.81      1407
   macro avg       0.75      0.72      0.73      1407
weighted avg       0.80      0.81      0.80      1407



### Random Forest

In [258]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf.fit(x_train, y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=1000,
                      n_jobs=None, oob_score=False, random_state=42, verbose=0,
                      warm_start=False)

In [396]:
pred= rf.predict(x_test)

score_rf=accuracy_score(y_test,pred.round())*100
score_rf

78.3226723525231

In [391]:
print ("\n Classification report : \n",classification_report(y_test,pred.round()))


 Classification report : 
               precision    recall  f1-score   support

           0       0.83      0.89      0.86      1038
           1       0.61      0.49      0.54       369

    accuracy                           0.78      1407
   macro avg       0.72      0.69      0.70      1407
weighted avg       0.77      0.78      0.78      1407



### XGBoost

In [361]:
import xgboost as xgb
from sklearn.metrics import f1_score, classification_report, accuracy_score
xgb_model = xgb.XGBClassifier(max_depth=2).fit(x_train, y_train) 
#, , learning_rate=0.1, colsample_bytree=.7, reg_alpha=4, objective='binary:logistic', eta=0.3, silent=1, subsample=0.8
xgb_prediction = xgb_model.predict(x_test)

In [398]:
accuracy = accuracy_score(y_test, xgb_prediction)
score_XG=accuracy*100
score_XG

80.59701492537313

In [390]:
XG_report=classification_report(y_test, xgb_prediction)
print(XG_report)

              precision    recall  f1-score   support

           0       0.84      0.91      0.87      1038
           1       0.67      0.52      0.58       369

    accuracy                           0.81      1407
   macro avg       0.75      0.71      0.73      1407
weighted avg       0.80      0.81      0.80      1407



### SVM

In [400]:
from sklearn import model_selection, svm
from sklearn.metrics import accuracy_score

SVM = svm.SVC(C=1.0, kernel='linear', degree=3, gamma='auto')
SVM.fit(x_train, y_train)

pred_SVM = SVM.predict(x_test)

score_svm=accuracy_score(y_test,pred_SVM)*100
score_svm

80.17057569296375

In [389]:
SVM_report=classification_report(y_test, pred_SVM)
print(SVM_report)

              precision    recall  f1-score   support

           0       0.82      0.93      0.87      1038
           1       0.70      0.43      0.53       369

    accuracy                           0.80      1407
   macro avg       0.76      0.68      0.70      1407
weighted avg       0.79      0.80      0.78      1407



### Final Conclusions

In [403]:
print("Accuracy Of Different Algorithms\n")
print("Logistic Regression:",score)
print("\nRandom Forest:",score_rf )
print("\nXG Boost:",score_XG)
print("\nSVM:",score_svm)


Accuracy Of Different Algorithms

Logistic Regression: 80.59701492537313

Random Forest: 78.3226723525231

XG Boost: 80.59701492537313

SVM: 80.17057569296375


## Logistic Regression performed better than other algorithms as it is a classification problem. XGboost also performed exactly similar to logistic regression. So Maximum Accuracy achieved was of 80.6%
